types/underscore/underscore-tests.ts

```diff
@@ -1744,9 +1744,20 @@ undefinedIdentityIterateeResult; // $ExpectType StringRecord
 
 // compact
 {
+    // lists
     _.compact(truthyFalsyList); // $ExpectType (string | number | true | object | Function | StringRecord | (() => void))[]
     _(truthyFalsyList).compact(); // $ExpectType (string | number | true | object | Function | StringRecord | (() => void))[]
     extractChainTypes(_.chain(truthyFalsyList).compact()); // $ExpectType ChainType<(string | number | true | object | Function | StringRecord | (() => void))[], string | number | true | object | Function | StringRecord | (() => void)>
+
+    // null
+    _.compact(null); // $ExpectType never[]
```

Comment:
What is `never[]` exactly? An array that can only be empty?

Reply:
I so badly want to be able to categorically say yes...but I can't. A `never[]` is an array that can only have contents of type `never`, so it is effectively an array that can only be empty, but it still ends up being an array that has a variable `length` property. The best type to describe an array that is now and forever will be empty would be the empty tuple type `[]`, which describes its `length` as a constant of `0`. All that being said, if one assigns a new array with no contents to a constant variable the inferred type is `never[]`, so it's at least not like I'm making up something completely new here.

![image](https://user-images.githubusercontent.com/8692126/88218576-13653200-cc15-11ea-9f71-ff4a6d953332.png)
![image](https://user-images.githubusercontent.com/8692126/88218618-1fe98a80-cc15-11ea-95ab-c8dbfd68d4d1.png)
![image](https://user-images.githubusercontent.com/8692126/88218659-31329700-cc15-11ea-97fa-0e0e37d4b77d.png)

If you'd like me to translate `never[]` into `[]`, it looks like `any[] extends never[]` is false but `never[] extends never[]` is true, so it looks like I could do such a translation safely.

Reply:
Well, after re-reading this, I guess I can categorically say that `never[]` is an array that can only be empty, I suppose I just can't say that it's the **best** array that can only be empty. Let me know what your thoughts are here whenever you have a moment 😄

Reply:
Interestingly one can assign both of these to a typed array without an assertion and then do stuff to the typed array, so I guess a non-fixed length is technically more accurate in that case. Also I admittedly kind of like the added descriptiveness that `never` adds, so I guess overall I'm in favor of leaving this as-is even if it wasn't going to add complication to do the translation.

```TS
const neverArray = [];
const emptyTuple: [] = [];
const neverStringArray: string[] = neverArray;
const emptyStringArray: string[] = emptyTuple;
neverStringArray.push('string');
emptyStringArray.push('string');
```

Reply:
Hmm, thinking about these tests more though, passing in something that can only be `undefined` isn't a really terribly interesting case; it might be better to test what happens when a `_.List<Truthies | _.AnyFalsy> | null | undefined` is passed in (trying it out it ends up doing the same thing as just a `_.List<Truthies | _.AnyFalsy>`).

Reply:
Given your explanation, the empty tuple seems perhaps slightly cleaner, but I can definitely be content with `never[]`. So that's enough "yes" as far as my approval is concerned.

As for the tests, it's up to you whether you want to do another push.

Reply:
I do think that a List + null + undefined type union is really the more realistic case here, so I did update the tests to check that instead. If you wouldn't mind taking one more look whenever is convenient, I'd appreciate it!